source/louisHelper.py

```diff
@@ -0,0 +1,72 @@
+#louisHelper.py
+#A part of NonVisual Desktop Access (NVDA)
+#This file is covered by the GNU General Public License.
+#See the file COPYING for more details.
+#Copyright (C) 2018 NV Access Limited, Babbage B.V.
+
+"""Helper module to ease communication to and from liblouis."""
+
+import louis
+from logHandler import log
+import config
+
+LOUIS_TO_NVDA_LOG_LEVELS = {
+	louis.LOG_ALL: log.DEBUG,
+	louis.LOG_DEBUG: log.DEBUG,
+	louis.LOG_INFO: log.INFO,
+	louis.LOG_WARN: log.WARNING,
+	louis.LOG_ERROR: log.ERROR,
+	louis.LOG_FATAL: log.ERROR,
+}
+
+@louis.LogCallback
+def louis_log(level, message):
+	if not _isDebug():
+		return
+	NVDALevel = LOUIS_TO_NVDA_LOG_LEVELS.get(level, log.DEBUG)
+	if not log.isEnabledFor(NVDALevel):
+		return
+	message = message.decode("ASCII")
+	codepath = "liblouis at internal log level %d" % level
+	log._log(NVDALevel, message, [], codepath=codepath)
+
+def _isDebug():
+	return config.conf["debugLog"]["louis"]
+
+def initialize():
+	# Register the liblouis logging callback.
+	louis.registerLogCallback(louis_log)
+	# Set the log level to all.
+	# The NVDA logging callback will filter messages appropriately,
+	# i.e. error messages will be logged at the error level.
+	louis.setLogLevel(louis.LOG_ALL)
```

Comment:
Are you sure there is not a performance decrease with this? If there are a lot of log calls in Liblouis at level debug,  there will be many switches from c to Python, even if the log callback Python function returns early. I certainly agree we should enable for warning and error though.

Reply:
I haven't experienced any noticeable performance quirks. I guess we can do either of 4 things:

1. Leave as is
2. Initialize liblouis debug at louis level DEBUG, in which case the number of message drops significantly (i.e. as far as I know, all messages involved with a translation that succeeds, are logged below level debug and thus won't be logged in that case.
3. Do not initialize louis logging at all when the louis flag in the configuration is False. This requires a restart of NVDA for changes to the louis flag to take effect.
4. Revert the config spec change. Instead, initialize louis logging at level_none, in which case nothing is logged at all. However... that only adds the possibility to change the logging level at runtime.

Reply:
Good to know you're not feeling any performance decrease. I'd still vote for going with option 2. Assuming there is no good reason why those messages below debug need to be easily accessed on any copy of NVDA.